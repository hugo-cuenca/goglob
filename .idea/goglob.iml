<?xml version="1.0" encoding="UTF-8"?>
<module type="CPP_MODULE" version="4">
  <component name="NewModuleRootManager">
    <content url="file://$MODULE_DIR$">
      <sourceFolder url="file://$MODULE_DIR$/goglob-common/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-go-ffi/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-proc-macro/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-serde/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-proc-macro-tests/src" isTestSource="false" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-proc-macro-tests/tests" isTestSource="true" />
      <sourceFolder url="file://$MODULE_DIR$/goglob-serde-tests/src" isTestSource="false" />
      <excludeFolder url="file://$MODULE_DIR$/target" />
    </content>
    <orderEntry type="inheritedJdk" />
    <orderEntry type="sourceFolder" forTests="false" />
  </component>
</module>