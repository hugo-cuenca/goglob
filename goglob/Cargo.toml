[package]
name = "goglob"
version = "0.2.0"
description = "Shell pattern matching similar to golang's path.Match"
authors = ["Hugo Cuenca <hugo.cuenca.arana@gmail.com>"]
readme = "../README.md"
homepage = "https://github.com/hugo-cuenca/goglob/"
repository = "https://github.com/hugo-cuenca/goglob/"
license = "BSD-3-Clause"
categories = ["filesystem"]
keywords = ["glob", "path", "go", "pattern", "regex"]
exclude = [
    ".idea/*",
]
edition = "2021"

[dependencies]
goglob-common = "0.2.0"
goglob-proc-macro = { version = "0.2.0", optional = true }

[features]
proc-macro = ["dep:goglob-proc-macro"]
serde = ["goglob-common/serde"]