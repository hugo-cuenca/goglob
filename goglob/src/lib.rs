//! Shell pattern matching similar to golang's `path.Match`.
//!
//! The pattern syntax is:
//!
//! ```text
//! pattern:
//!     { term }
//! term:
//!     '*'         matches any sequence of non-/ characters
//!     '?'         matches any single non-/ character
//!     '[' [ '^' ] { character-range } ']'
//!                 character class (must be non-empty)
//!     c           matches character c (c != '*', '?', '\\', '[')
//!     '\\' c      matches character c
//!
//! character-range:
//!     c           matches character c (c != '\\', '-', ']')
//!     '\\' c      matches character c
//!     lo '-' hi   matches character c for lo <= c <= hi
//! ```
//!
//! Match requires pattern to match all of name, not just a substring.
//!
//! Use [`GlobPattern::new(pattern)`][GlobPattern::new] to construct a new instance.
//!
//! # Features
//! * `proc-macro`: allows using the `glob!("<PATTERN>")` procedural macro (see
//!   [glob!()][glob]).
//! * `serde`: enables serde deserialization of string patterns.
//!
//! # License
//! `BSD-3-Clause`.
//!
//! Based in [Go 1.18's `path.Match`](https://cs.opensource.google/go/go/+/refs/tags/go1.18:src/path/match.go;l=38)
//! (available under the [`BSD-3-Clause` license](https://cs.opensource.google/go/go/+/refs/tags/go1.18:LICENSE))
//! which contains the following comment:
//!
//! ```go
//! // Copyright 2010 The Go Authors. All rights reserved.
//! // Use of this source code is governed by a BSD-style
//! // license that can be found in the LICENSE file.
//! ```

pub use goglob_common::error;
pub use goglob_common::Result;

pub use goglob_common::GlobPattern;

#[cfg(feature = "proc-macro")]
pub use goglob_proc_macro::*;

#[cfg(feature = "proc-macro")]
#[doc(hidden)]
pub mod internal {
    pub use goglob_common::*;
}
