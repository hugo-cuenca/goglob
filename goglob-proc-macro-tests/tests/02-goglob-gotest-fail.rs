use goglob::{glob, GlobPattern};

const PATTERN_01: GlobPattern = glob!("[]a]");
const PATTERN_02: GlobPattern = glob!("[-]");
const PATTERN_03: GlobPattern = glob!("[x-]");
const PATTERN_04: GlobPattern = glob!("[-x]");
const PATTERN_05: GlobPattern = glob!("\\");
const PATTERN_06: GlobPattern = glob!("[a-b-c]");
const PATTERN_07: GlobPattern = glob!("[");
const PATTERN_08: GlobPattern = glob!("[^");
const PATTERN_09: GlobPattern = glob!("[^bc");
const PATTERN_10: GlobPattern = glob!("a[");
const PATTERN_11: GlobPattern = glob!("a/b[");

fn main() {}