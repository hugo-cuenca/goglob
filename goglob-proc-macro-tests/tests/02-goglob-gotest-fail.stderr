error: pattern malformed: special character ] at 1 not escaped with '\'
 --> tests/02-goglob-gotest-fail.rs:3:39
  |
3 | const PATTERN_01: GlobPattern = glob!("[]a]");
  |                                       ^^^^^^

error: pattern malformed: special character - at 1 not escaped with '\'
 --> tests/02-goglob-gotest-fail.rs:4:39
  |
4 | const PATTERN_02: GlobPattern = glob!("[-]");
  |                                       ^^^^^

error: pattern malformed: special character ] at 3 not escaped with '\'
 --> tests/02-goglob-gotest-fail.rs:5:39
  |
5 | const PATTERN_03: GlobPattern = glob!("[x-]");
  |                                       ^^^^^^

error: pattern malformed: special character - at 1 not escaped with '\'
 --> tests/02-goglob-gotest-fail.rs:6:39
  |
6 | const PATTERN_04: GlobPattern = glob!("[-x]");
  |                                       ^^^^^^

error: pattern malformed: illegal use of '\' at 0: end of pattern
 --> tests/02-goglob-gotest-fail.rs:7:39
  |
7 | const PATTERN_05: GlobPattern = glob!("\\");
  |                                       ^^^^

error: pattern malformed: special character - at 4 not escaped with '\'
 --> tests/02-goglob-gotest-fail.rs:8:39
  |
8 | const PATTERN_06: GlobPattern = glob!("[a-b-c]");
  |                                       ^^^^^^^^^

error: pattern malformed: character class opened with '[' at 0 isn't closed
 --> tests/02-goglob-gotest-fail.rs:9:39
  |
9 | const PATTERN_07: GlobPattern = glob!("[");
  |                                       ^^^

error: pattern malformed: character class opened with '[' at 0 isn't closed
  --> tests/02-goglob-gotest-fail.rs:10:39
   |
10 | const PATTERN_08: GlobPattern = glob!("[^");
   |                                       ^^^^

error: pattern malformed: character class opened with '[' at 0 isn't closed
  --> tests/02-goglob-gotest-fail.rs:11:39
   |
11 | const PATTERN_09: GlobPattern = glob!("[^bc");
   |                                       ^^^^^^

error: pattern malformed: character class opened with '[' at 1 isn't closed
  --> tests/02-goglob-gotest-fail.rs:12:39
   |
12 | const PATTERN_10: GlobPattern = glob!("a[");
   |                                       ^^^^

error: pattern malformed: character class opened with '[' at 3 isn't closed
  --> tests/02-goglob-gotest-fail.rs:13:39
   |
13 | const PATTERN_11: GlobPattern = glob!("a/b[");
   |                                       ^^^^^^
